c7n/resources/ec2.py

```diff
@@ -898,6 +898,139 @@ def process_resource_set(self, client, resources):
+    'disabled'|'disabling'|'enabled'|'pending'
+    (https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ec2.html#EC2.Client.describe_instances)
+
+    :Example:
+
+    .. code-block:: yaml
+
+        policies:
+          - name: ec2-detailed-monitoring-activation
+            resource: ec2
+            filters:
+              - type: monitoring-status
+                key: State
+                value: disabled
+            actions:
+              - type: enable-monitoring
+    References
+
+     https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-cloudwatch-new.html
+    """
+    valid_values = ['disabled', 'disabling', 'enabled', 'pending']
+
+    schema = type_schema(
+        'monitoring-status',
+        rinherit=ValueFilter.schema,
+        value={'enum': valid_values},
+        required=('value',))
+    permissions = list('ec2:DescirbeInstances')
+
+    def process(self, resources, event=None):
+        client = utils.local_session(
+            self.manager.session_factory).client('ec2')
+        return [r for r in resources
+                if self.check_monitoring_status(client, r, self.data.get('value'))]
+
+    def check_monitoring_status(self, client, inst, value):
+        resp = client.describe_instances(
+            InstanceIds=[inst['InstanceId']]
+        )
+        if (resp['Reservations'][0]['Instances'][0]['Monitoring']['State'] == value and
+                resp['Reservations'][0]['Instances'][0]['State']['Name']
+                not in ['terminated', 'shutting-down']):
+            return True
+        return False
+
+
+@actions.register('enable-monitoring')
+class MonitorInstances(BaseAction, StateTransitionFilter):
+    """Action on EC2 Instances to enable detailed monitoring
+
+    The differents states of detailed monitoring status are :
+    'disabled'|'disabling'|'enabled'|'pending'
+    (https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ec2.html#EC2.Client.describe_instances)
+
+    :Example:
+
+    .. code-block:: yaml
+
+        policies:
+          - name: ec2-detailed-monitoring-activation
+            resource: ec2
+            filters:
+              - type: monitoring-status
+                key: State
+                value: disabled
+            actions:
+              - type: enable-monitoring
+    References
+
+     https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-cloudwatch-new.html
+    """
+    schema = type_schema('enable-monitoring')
+    permissions = list('ec2:MonitorInstances')
+
+    def process(self, resources, event=None):
+        if not len(resources):
+            return
+        client = utils.local_session(
+            self.manager.session_factory).client('ec2')
+        self.enable_monitoring(client, resources)
+
+    def enable_monitoring(self, client, resources):
+        client.monitor_instances(
+            InstanceIds=[inst['InstanceId'] for inst in resources]
+        )
+
+
+@actions.register('disable-monitoring')
```

Comment:
we no longer do enable/disable, we do a single action `set` that can produce either state based on parameters, iee

```
- type: set-monitoring
   state: enabled
```